src/Init/System/IO.lean

```diff
@@ -473,22 +473,11 @@ partial def Handle.readBinToEnd (h : Handle) : IO ByteArray := do
       loop (acc ++ buf)
   loop ByteArray.empty
 
-partial def Handle.readToEnd (h : Handle) : IO String := do
-  let rec loop (s : String) := do
-    let line ← h.getLine
-    if line.isEmpty then
-      return s
-    else
-      loop (s ++ line)
-  loop ""
-
-def readBinFile (fname : FilePath) : IO ByteArray := do
-  let h ← Handle.mk fname Mode.read
-  h.readBinToEnd
-
-def readFile (fname : FilePath) : IO String := do
-  let h ← Handle.mk fname Mode.read
-  h.readToEnd
+def Handle.readToEnd (h : Handle) : IO String := do
+  let data ← h.readBinToEnd
+  match String.fromUTF8? data with
+  | some s => return s
+  | none => throw <| .userError s!"Tried to read from handle containing non UTF-8 data."
```

Comment:
A feature request for another time; `fromUTF8?` should return an `Except` that indicates where the non-utf8 character was, even if just as a string error message.